config.txt
# Graphics Settings
Bloom                : true
Phong                : true
Shadows              : true
Soft Shadows         : true
Shadow Resolution    : 1024
Stereo 3D            : false
V-Sync               : false
FSAA                 : true
FSAA Samples         : 8
Target FPS           : 60
First Person FOV     : 120
Third Person FOV     : 80
Frame Width          : 1024
Frame Height         : 768
Frame X              : 0
Frame Y              : 0
Center Frame         : true
Frame Maximized      : false
Save Frame State     : true

# Overlay Default Visibility
Server Speed         : true
Foul Overlay         : true
Field Overlay        : true
Number of Players    : true
Player IDs           : true

# Networking Settings
Auto-Connect         : true
Auto-Connect Delay   : 1000
Default Server       : localhost:60001
Drawing Port         : 32769
Monitor Step         : 0.04
Network Buffer       : false
# To store more servers, just insert more "Server" lines
Server               : localhost:60001

# General Settings
Record Logfiles      : false
Logfile Directory    : 
# Possible values for "Look and Feel" are: "system", "darcula", "intellij", "solarized_light" and "solarized_dark"
Look and Feel        : system

# Team Colors
Team Color           : <Right>:0xff2626
Team Color           : <Left>:0x2626ff


client/config.txt
Graphics Settings:
Bloom                : true
Phong                : true
Shadows              : true
Soft Shadows         : true
Shadow Resolution    : 1024
Stereo 3D            : false
V-Sync               : false
FSAA                 : true
FSAA Samples         : 8
Target FPS           : 60
First Person FOV     : 120
Third Person FOV     : 80
Frame Width          : 1024
Frame Height         : 768
Frame X              : 0
Frame Y              : 0
Center Frame         : true
Frame Maximized      : false
Save Frame State     : true

Overlay Default Visibility:
Server Speed         : true
Foul Overlay         : true
Field Overlay        : true
Number of Players    : true
Player IDs           : true

Networking Settings:
Auto-Connect         : true
Auto-Connect Delay   : 1000
Server Hosts         : localhost
Server Port          : 3200
Drawing Port         : 32769

General Settings:
Record Logfiles      : false
Logfile Directory    : 

Team Colors:
<Right>              : 0xff2626
<Left>               : 0x2626ff



CHANGELOG.md
2.1.1 (July 2, 2026)
------------------------------
* fix rendering of penalty/goalie area lines
* fix ball circle rendering to also work with rcssservermj

2.1.0 (June 28, 2026)
------------------------------
* add compatibility for rcssservermj (newly supported robots: T1, K1)
* use a new procedurally generated field texture
* add support for loading STL models
* exit application on ctrl-Q
* fix requested target FPS value from config being ignored
* make ball tracking speed independent of FPS
* respect setVisible attribute of geometry nodes
* apply correct material for SingleMaterialNodes
* update default monitor port to 60001

2.0.1 (April 17, 2025)
------------------------------
* add support for reading log files with a .tgz or .tbz2 suffix
* fix reading logs from zip files
* fix "Step back" and "Step forward" buttons in the log player

2.0.0 (October 6, 2023)
------------------------------
* update minimum required Java version to 17
* require strings sent via the drawing protocol to be UTF-8 encoded
* update JOGL to 2.4.0 (adds support for macOS Ventura)
* move entire build logic to Gradle (use `./gradlew binDir` instead of
  the old build script)
* add a build task to generate a macOS application bundle
  (`./gradlew macOSApp`)
* support running without a present config file
* make the team color table in the configuration window scrollable
* fix a NullPointerException when stopping the log player
* fix an IndexOutOfBoundsException when the connection ends unexpectedly

1.8.5 (July 2, 2023)
------------------------------
* add timer for penalties in RCSSServer3D 0.7.6

1.8.4 (April 29, 2023)
------------------------------
* increase tracking speed of camera when tracking the ball

1.8.3 (March 18, 2023)
------------------------------
* add support for ball holding foul introduced with RCSSServer3D 0.7.6
* add optional buffering mechanism for bad network connections

1.8.2 (July 14, 2022)
------------------------------
* assign default team color based on team name

1.8.1 (April 7, 2022)
------------------------------
* fixed crashing RoboViz due to an invalid regex ([#52](https://github.com/magmaOffenburg/RoboViz/issues/52))
* migrated logging to Log4j ([#124](https://github.com/magmaOffenburg/RoboViz/pull/124))

1.8.0 (December 30, 2021)
------------------------------
* included the [darklaf](https://github.com/weisJ/darklaf) library for better look and feels ([#110](https://github.com/magmaOffenburg/RoboViz/pull/110))
* added "Look and Feel" to "General Settings" in `config.txt` - possible values are:
    * `system` (default)
    * `darcula`
    * `intellij`
    * `solarized_dark`
    * `solarized_light`
* reimplemented the RoboViz GUI in Kotlin to better separate GUI and Rendering ([#117](https://github.com/magmaOffenburg/RoboViz/pull/117))
    * settings window can be opened via menu (graphics settings can be changed without a restart)
    * default team colors can be overwritten
    * switch between live and log mode via menu ([#55](https://github.com/magmaOffenburg/RoboViz/issues/55))
    * fixed playback speed text field not selectable on Linux ([#31](https://github.com/magmaOffenburg/RoboViz/issues/31))
* configuration improvements ([#112](https://github.com/magmaOffenburg/RoboViz/pull/112))
    * Commented out lines are kept on save
    * New syntax for storing multiple servers and team colors
* made camera movement speed relative to the distance from the ground
* pass mode timer update ([#114](https://github.com/magmaOffenburg/RoboViz/pull/114))

1.7.0 (February 21, 2020)
------------------------------
* added "Connection -> Connect to..." ([#105](https://github.com/magmaOffenburg/RoboViz/issues/105))
* switched to Gradle for build / dependency management ([#106](https://github.com/magmaOffenburg/RoboViz/issues/106))
    * simplified build scripts (there's only `build.sh` for Linux / Mac and `build.bat` for Windows)
    * builds are now universal (same `bin` folder works on Windows, Linux and Mac)
    * updated jogl to 2.3.2 (fixes Mac compatibility - [#93](https://github.com/magmaOffenburg/RoboViz/issues/93))
* fixed "Auto Connect" causing unresponsiveness in the UI 
* allow more precise zooming with Shift + Mouse Wheel

1.6.1 (June 17, 2019)
------------------------------
* added support for the rectangular goals in rcssserver 0.7.2
* updated the pass mode duration for rcssserver 0.7.2

1.6.0 (May 1, 2019)
------------------------------
* updated the required Java version to 1.8
* added a fade-out effect to the ball circle in pass mode etc.
* added a workaround for goals sometimes being missed ([#104](https://github.com/magmaOffenburg/RoboViz/issues/104))
* added an overlay text and a gray-out effect when in `GameOver`
* changed the default config to enable all overlays as in competitions
* removed red cards from the foul overlay to avoid confusion
* removed the "Fouls:" headline from the foul overlay

1.5.0 (April 27, 2019)
------------------------------
* added differently colored goalie jerseys ([#102](https://github.com/magmaOffenburg/RoboViz/issues/102))
* added a player tracking camera
* added support for the new self collision foul
* added a circle around the ball for pass / free kick / corner kick / kick in
* added a timer for the scoring cooldown after pass mode 
* changed the time formatting to `mm:ss`
* more gracefully handle `NaN` values sent by the server ([#99](https://github.com/magmaOffenburg/RoboViz/issues/99))
* fixed hangs when switching between servers in Connection UI ([#100](https://github.com/magmaOffenburg/RoboViz/issues/100))
* fixed `config.txt` changes made while RoboViz is running being lost on save

1.4.0 (April 12, 2018)
------------------------------
* added a menu bar with Connection / Server / View / Camera menus
* allowed specifying multiple servers in `config.txt` (comma-separated)
* allowed switching between different servers without restarting RoboViz

src/main/kotlin/org/magmaoffenburg/roboviz/configuration/Config.kt
package org.magmaoffenburg.roboviz.configuration

import org.apache.logging.log4j.kotlin.logger
import org.magmaoffenburg.roboviz.etc.ConfigChangeListener
import java.awt.Color
import java.io.File

/**
 * To support multiple Team colors and servers without breaking existing
 * Config files, "Server :" and "Team Color" are SuperKeys. The ConfigParser can
 * get all lines starting with a SuperKey. The raw SuperKeys are trimmed to only
 * contain one space.
 */
class Config(args: Array<String>) {

    private val logger = logger()

    private val localPath = "config.txt"
    private val globalPath = System.getProperty("user.home") + "/.roboviz/config.txt"

    private val parser = ConfigParser()
    val filePath = if (File(globalPath).exists() || !File(localPath).exists()) globalPath else localPath

    private val configChangeListeners = arrayListOf<ConfigChangeListener>()

    init {
        parser.parseArgs(args)
        if (filePathExists()) {
            parser.parseFile(filePath)
        }

        try {
            read()
        } catch (ex: Exception) {
            logger.error("Error reading parsed values. The configuration file might be corrupt or incompatible with this version of RoboViz, try resetting it.")
        }
    }

    fun filePathExists() = File(filePath).exists()

    object General {
        var recordLogs = false
        var logfileDirectory = ""
        var logReplayFile = ""
        var lookAndFeel = "system"
        var drawingFilter = ".*"
    }

    object Graphics {
        var useBloom = true
        var usePhong = true
        var useShadows = true
        var useSoftShadows = true
        var shadowResolution = 1024
        var useStereo = false
        var useVsync = false
        var useFsaa = true
        var fsaaSamples = 8
        var targetFPS = 60
        var firstPersonFOV = 120
        var thirdPersonFOV = 80

        var frameWidth = 1024
        var frameHeight = 768
        var frameX = 0
        var frameY = 0

        var centerFrame = true
        var isMaximized = false
        var saveFrameState = true
    }

    object Networking {
        var autoConnect = true
        var autoConnectDelay = 1000
        var listenPort = 32769

        var servers = arrayListOf<Pair<String, Int>>()
        var defaultServerHost = "localhost"
        var defaultServerPort = 60001

        var currentHost = defaultServerHost
        var currentPort = defaultServerPort

        var monitorStep = 0.04
        var useBuffer = false
    }

    object OverlayVisibility {
        var serverSpeed = true
        var foulOverlay = true
        var fieldOverlay = true
        var numberOfPlayers = true
        var playerIDs = true
    }

    object TeamColors {
        val byTeamNames = hashMapOf<String, Color>()
        val defaultLeft = Color(0x2626ff)
        val defaultRight = Color(0xff2626)

        fun getLeftColor(teamNameLeft: String?, teamNameRight: String?) =
            byTeamNames[teamNameLeft] ?: if (teamNameLeft != null && teamNameRight != null) {
                getOrderedColor(teamNameLeft, teamNameRight, true)
            } else {
                byTeamNames.getOrDefault(defaultLeftTeamName, defaultLeft)
            }

        fun getRightColor(teamNameLeft: String?, teamNameRight: String?) =
            byTeamNames[teamNameRight] ?: if (teamNameLeft != null && teamNameRight != null) {
                getOrderedColor(teamNameLeft, teamNameRight, false)
            } else {
                byTeamNames.getOrDefault(defaultRightTeamName, defaultRight)
            }

        private fun getOrderedColor(leftTeam: String, rightTeam: String, left: Boolean): Color {
            val switchColors = leftTeam < rightTeam
            return if ((left && !switchColors) || (!left && switchColors)) {
                byTeamNames.getOrDefault(defaultLeftTeamName, defaultLeft)
            } else {
                byTeamNames.getOrDefault(defaultRightTeamName, defaultRight)
            }
        }
    }

    companion object {
        const val defaultLeftTeamName = "<Left>"
        const val defaultRightTeamName = "<Right>"
    }

    /**
     * set object variables to values from parser
     */
    fun read() {
        // General
        parser.getValue("Record Logfiles")?.let { General.recordLogs = it.toBoolean() }
        parser.getValue("Logfile Directory")?.let { General.logfileDirectory = it }
        parser.getValue("Look and Feel")?.let { General.lookAndFeel = it }

        // Graphics
        parser.getValue("Bloom")?.let { Graphics.useBloom = it.toBoolean() }
        parser.getValue("Phong")?.let { Graphics.usePhong = it.toBoolean() }
        parser.getValue("Shadows")?.let { Graphics.useShadows = it.toBoolean() }
        parser.getValue("Soft Shadows")?.let { Graphics.useSoftShadows = it.toBoolean() }
        parser.getValue("Shadow Resolution")?.let { Graphics.shadowResolution = it.toInt() }
        parser.getValue("Stereo 3D")?.let { Graphics.useStereo = it.toBoolean() }
        parser.getValue("V-Sync")?.let { Graphics.useVsync = it.toBoolean() }
        parser.getValue("FSAA")?.let { Graphics.useFsaa = it.toBoolean() }
        parser.getValue("FSAA Samples")?.let { Graphics.fsaaSamples = it.toInt() }
        parser.getValue("Target FPS")?.let { Graphics.targetFPS = it.toInt() }
        parser.getValue("First Person FOV")?.let { Graphics.firstPersonFOV = it.toInt() }
        parser.getValue("Third Person FOV")?.let { Graphics.thirdPersonFOV = it.toInt() }

        parser.getValue("Frame Width")?.let { Graphics.frameWidth = it.toInt() }
        parser.getValue("Frame Height")?.let { Graphics.frameHeight = it.toInt() }
        parser.getValue("Frame X")?.let { Graphics.frameX = it.toInt() }
        parser.getValue("Frame Y")?.let { Graphics.frameY = it.toInt() }
        parser.getValue("Center Frame")?.let { Graphics.centerFrame = it.toBoolean() }
        parser.getValue("Frame Maximized")?.let { Graphics.isMaximized = it.toBoolean() }
        parser.getValue("Save Frame State")?.let { Graphics.saveFrameState = it.toBoolean() }

        // Networking
        parser.getValue("Auto-Connect")?.let { Networking.autoConnect = it.toBoolean() }
        parser.getValue("Auto-Connect Delay")?.let { Networking.autoConnectDelay = it.toInt() }
        parser.getValue("Drawing Port")?.let { Networking.listenPort = it.toInt() }
        parser.getValue("Monitor Step")?.let { Networking.monitorStep = it.toDouble() }
        parser.getValue("Network Buffer")?.let { Networking.useBuffer = it.toBoolean() }

        parser.getValuePairList("Server").forEach {
            Networking.servers.add(Pair(it.first, it.second.toInt()))
        }
        parser.getValue("Default Server")?.let { Networking.defaultServerHost = it.substringBefore(":") }
        parser.getValue("Default Server")?.let { Networking.defaultServerPort = it.substringAfter(":").toInt() }

        // OverlayVisibility
        parser.getValue("Server Speed")?.let { OverlayVisibility.serverSpeed = it.toBoolean() }
        parser.getValue("Foul Overlay")?.let { OverlayVisibility.foulOverlay = it.toBoolean() }
        parser.getValue("Field Overlay")?.let { OverlayVisibility.fieldOverlay = it.toBoolean() }
        parser.getValue("Number of Players")?.let { OverlayVisibility.numberOfPlayers = it.toBoolean() }
        parser.getValue("Player IDs")?.let { OverlayVisibility.playerIDs = it.toBoolean() }

        // TeamColors
        parser.getValuePairList("Team Color").forEach {
            TeamColors.byTeamNames[it.first] = Color(Integer.decode(it.second))
        }
        TeamColors.byTeamNames.putIfAbsent(defaultLeftTeamName, TeamColors.defaultLeft)
        TeamColors.byTeamNames.putIfAbsent(defaultRightTeamName, TeamColors.defaultRight)

        // Args
        Networking.currentHost = parser.argsList.firstOrNull {
            it.first == "serverHost"
        }?.second ?: Networking.defaultServerHost
        Networking.currentPort = parser.argsList.firstOrNull {
            it.first == "serverPort"
        }?.second?.toInt() ?: Networking.defaultServerPort
        General.logReplayFile = parser.argsList.firstOrNull {
            it.first == "logFile"
        }?.second ?: ""
        General.drawingFilter = parser.argsList.firstOrNull {
            it.first == "drawingFilter"
        }?.second ?: ".*"
    }

    /**
     *  set parser list to object variable values and write config to disk
     *  @param initConfig whether to initialize the config file if it does not exist yet
     */
    fun write(initConfig: Boolean = false) {
        // general
        parser.setValue("Record Logfiles", General.recordLogs.toString())
        parser.setValue("Logfile Directory", General.logfileDirectory)
        parser.setValue("Look and Feel", General.lookAndFeel)

        // Graphics
        parser.setValue("Bloom", Graphics.useBloom.toString())
        parser.setValue("Phong", Graphics.usePhong.toString())
        parser.setValue("Shadows", Graphics.useShadows.toString())
        parser.setValue("Soft Shadows", Graphics.useSoftShadows.toString())
        parser.setValue("Shadow Resolution", Graphics.shadowResolution.toString())
        parser.setValue("Stereo 3D", Graphics.useStereo.toString())
        parser.setValue("V-Sync", Graphics.useVsync.toString())
        parser.setValue("FSAA", Graphics.useFsaa.toString())
        parser.setValue("FSAA Samples", Graphics.fsaaSamples.toString())
        parser.setValue("Target FPS", Graphics.targetFPS.toString())
        parser.setValue("First Person FOV", Graphics.firstPersonFOV.toString())
        parser.setValue("Third Person FOV", Graphics.thirdPersonFOV.toString())

        parser.setValue("Frame Width", Graphics.frameWidth.toString())
        parser.setValue("Frame Height", Graphics.frameHeight.toString())
        parser.setValue("Frame X", Graphics.frameX.toString())
        parser.setValue("Frame Y", Graphics.frameY.toString())
        parser.setValue("Center Frame", Graphics.centerFrame.toString())
        parser.setValue("Frame Maximized", Graphics.isMaximized.toString())
        parser.setValue("Save Frame State", Graphics.saveFrameState.toString())

        // Networking
        parser.setValue("Auto-Connect", Networking.autoConnect.toString())
        parser.setValue("Auto-Connect Delay", Networking.autoConnectDelay.toString())
        parser.setValue("Drawing Port", Networking.listenPort.toString())

        parser.setValuePairList("Server", Networking.servers.map { Pair(it.first, it.second.toString()) })
        parser.setValue("Default Server", "${Networking.defaultServerHost}:${Networking.defaultServerPort}")

        parser.setValue("Monitor Step", Networking.monitorStep.toString())
        parser.setValue("Network Buffer", Networking.useBuffer.toString())

        // OverlayVisibility
        parser.setValue("Server Speed", OverlayVisibility.serverSpeed.toString())
        parser.setValue("Foul Overlay", OverlayVisibility.foulOverlay.toString())
        parser.setValue("Field Overlay", OverlayVisibility.fieldOverlay.toString())
        parser.setValue("Number of Players", OverlayVisibility.numberOfPlayers.toString())
        parser.setValue("Player IDs", OverlayVisibility.playerIDs.toString())

        // team colors
        parser.setValuePairList(
            "Team Color",
            TeamColors.byTeamNames.map { Pair(it.key, "0x${Integer.toHexString(it.value.rgb and 0xFFFFFF)}") })

        if (initConfig || filePathExists()) {
            try {
                parser.writeFile(filePath)
            } catch (e: ArrayIndexOutOfBoundsException) {
                logger.error("Error while trying to save the config.", e)
            }
        }
    }

    fun configChanged() {
        logger.info { "Config changed" }

        configChangeListeners.forEach {
            it.onConfigChanged()
        }
    }

    fun addConfigChangedListener(listener: ConfigChangeListener) {
        configChangeListeners.add(listener)
    }

    @Suppress("unused")
    fun removeConfigChangedListener(listener: ConfigChangeListener) {
        configChangeListeners.remove(listener)
    }

}
